### PR TITLE
fix: guard BigNumber usage for strict-mode (v10+)

### PR DESCRIPTION
**Pull request by simcheolhwan**
## Summary

- Harden every BigNumber call site against empty-string and other invalid inputs so the library keeps working when consumers run on `bignumber.js@>=10`, where invalid input throws by default (v10 removed the opt-in `DEBUG` flag; v11 keeps the same default behind a new `STRICT` toggle).
- Replace `?? "0"` with `|| "0"` (nullish coalescing lets `""` through) and add `|| 0` guards to BigNumber constructors and chained methods (`.plus()`, `.times()`, `.lt()`, etc., which internally construct BigNumbers from their arguments).
- Drop empty gas-price entries at the source (`fetchGasPrices`) with a `console.error`, instead of letting them flow downstream as a silent zero-fee option that would let users submit unpayable transactions.
- Refactor `isInsufficientBalance` so unparseable `quantity` (e.g. `"abc"`) returns `false` via the `toBaseUnit` empty-string fallback instead of throwing in `BigNumber(quantity).isZero()`.
- Enable `BigNumber.DEBUG = true` permanently via a new `vitest.setup.ts` so every existing test acts as a strict-mode regression check.
- Add regression tests for empty-string inputs in `amountValidation` and Skip fee filtering.
- Document the convention in `AGENTS.md` (final-fallback placement, full method list, the `@initia/utils` empty-string source, and the test-time guarantee).

## Test plan

- [ ] `pnpm typecheck` passes
- [ ] `pnpm test` passes (508/508 with `BigNumber.DEBUG = true` in setup)
- [ ] `pnpm lint` passes
- [ ] Manual: open the widget, walk through Send / Bridge / Deposit / Withdraw flows with a connected wallet; verify quantity input, balance display, fee dropdown, and tx preview behave unchanged
- [ ] Manual: confirm no regressions on the Activity tab amount rendering

---

> [!NOTE]
> **Medium Risk**
> Touches many transaction-critical paths (bridge/deposit/send fee math, balance validation, route querying) to change how numeric inputs are parsed and defaulted, which could subtly affect fee/amount calculations. Added tests and strict-mode enforcement reduce regression risk but broad surface area warrants careful review.
> 
> **Overview**
> Makes BigNumber usage resilient to v10+ strict-mode by introducing `parseQuantity()` and routing user/remote/localStorage numeric strings through it, plus replacing many `??` fallbacks with `||` (and adding `|| 0`/`|| "0"`) to prevent empty-string inputs from throwing.
> 
> Updates fee/route handling to avoid unsafe defaults: L1 gas prices are normalized via `normalizeL1GasPriceEntry()`, empty/whitespace entries are dropped with an error log, and `fetchGasPrices()` now throws when all entries are invalid; slippage validation is centralized in `getSlippageMessage()` and persisted slippage is normalized before replay. Adds regression tests for parsing, slippage persistence, Skip query key normalization, fee warning behavior, and enables strict-mode checking in Vitest via `vitest.setup.ts`, plus documents the fallback rules in `AGENTS.md`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f5257aa36b9012ee7601485f50270768c27d1983. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added a safe quantity parser utility to normalize and validate user numeric input.

* **Bug Fixes**
  * Made balance, fee, quantity and sorting calculations resilient to missing/empty numeric inputs across bridge, deposit, transaction, portfolio and wallet flows.
  * Ignored empty/whitespace fee and gas-price entries to prevent erroneous fee/sorting/display outcomes.

* **Documentation**
  * Added guidance on strict handling of numeric inputs and required final fallbacks.

* **Tests**
  * Added amount-validation tests for empty/unparseable inputs and enabled stricter test-time numeric validation.

### CODE REVIEW

File: packages/interwovenkit-react/src/data/fee.ts [hunk 1]
```diff
@@ -22,10 +22,18 @@ export async function fetchGasPrices(chain: NormalizedChain) {
...
+        // Drop entries with empty/missing amounts so downstream fee selection
+        // can't silently treat them as a zero-fee option (which would let
+        // transactions through with an unpayable fee).
+        if (!amount) {
+          // eslint-disable-next-line no-console
+          console.error(`Gas price entry has empty amount for ${denom}`)
+          return []
+        }
         const multiplier = denom === "uinit" ? 1 : DEFAULT_GAS_PRICE_MULTIPLIER
         const price = BigNumber(amount).times(multiplier).toFixed(18)
```
[the remark this answers is not in the document]
Comment (simcheolhwan) [reply packages/interwovenkit-react/src/data/fee.ts]: Trimmed `amount` before the empty check in 1f5dd4c so whitespace-only values get dropped too. Skipped the `|| 0` fallback at the `BigNumber()` call though — after the `if (!trimmed) return []` guard the value is provably non-empty, and the explicit design here is to drop bad entries rather than coerce them to a zero fee (the surrounding comment calls that out). `|| 0` wouldn't help against non-numeric strings like `"abc"` either, so it'd be dead code at that position.

File: packages/interwovenkit-react/src/data/fee.ts [hunk 2]
```diff
@@ -22,10 +22,19 @@ export async function fetchGasPrices(chain: NormalizedChain) {
       .json<{ gas_prices: Coin[] }>()
     return [...gas_prices]
       .sort(descend(({ denom }) => denom === "uinit"))
-      .map(({ denom, amount }) => {
+      .flatMap(({ denom, amount }) => {
+        // Drop entries with empty/whitespace-only amounts so downstream fee
+        // selection can't silently treat them as a zero-fee option (which
+        // would let transactions through with an unpayable fee).
+        const trimmed = amount.trim()
```
Comment (tansawit) [inline packages/interwovenkit-react/src/data/fee.ts]: Can we normalize before trimming here? `amount.trim()` still throws if the API sends a missing/null amount, while the comment and PR summary say empty/missing gas-price entries are dropped. `const trimmed = (amount || "").trim()` would keep the intended drop path.

File: packages/interwovenkit-react/vitest.setup.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import BigNumber from "bignumber.js"
+
+// Enable strict mode so every test acts as a regression check against
+// re-introducing empty-string or otherwise invalid BigNumber inputs.
+// Mirrors the default behavior consumers get on bignumber.js@>=10.
+BigNumber.DEBUG = true
```
Comment (tansawit) [inline packages/interwovenkit-react/vitest.setup.ts]: This is v9-only from a typing perspective. The peer dep allows `bignumber.js >=9` and this PR targets v10+, but v10.0.2 has no `DEBUG` in its declarations and v11.1.1 exposes `STRICT` via config instead; resolving v10/v11 makes `typecheck` fail here with TS2339.

File: packages/interwovenkit-react/src/data/fee.ts (hunk 2 above)
Comment (simcheolhwan) [reply packages/interwovenkit-react/src/data/fee.ts]: Good catch — switched to `(amount || "").trim()` in 57edd03 so missing/null amounts now flow through the same drop path instead of throwing on `.trim()`.

File: packages/interwovenkit-react/vitest.setup.ts (hunk 3 above)
Comment (simcheolhwan) [reply packages/interwovenkit-react/vitest.setup.ts]: Right — the static `DEBUG` only exists in v9's `.d.ts`. Replaced the assignment with `(BigNumber as unknown as { DEBUG?: boolean }).DEBUG = true` in 57edd03 so it type-checks across v9/v10/v11. v9.3.1 is the currently installed resolution; v10+ already throws on invalid input by default, and v11 reaches the same default via `BigNumber.config({ STRICT: true })`, so the assignment is a no-op on v10+ but stays safe.

File: packages/interwovenkit-react/src/components/form/QuantityInput.tsx [hunk 4]
```diff
@@ -25,7 +25,7 @@ const QuantityInput = ({ balance, decimals, className }: Props) => {
   const rules = {
     required: "Enter amount",
     validate: (quantity: string) => {
-      if (BigNumber(quantity).isZero()) {
+      if (BigNumber(quantity || 0).isZero()) {
```
Comment (evilpeach) [inline packages/interwovenkit-react/src/components/form/QuantityInput.tsx]: Based on this video, `QuantityInput` should treat invalid numeric input like "." as invalid and keep Confirm disabled.

https://github.com/user-attachments/assets/891a57a4-234b-4190-a035-d8668831ac26

Comment (simcheolhwan) [reply packages/interwovenkit-react/src/components/form/QuantityInput.tsx]: Fixed in cf51ca0. Added a `parseQuantity` helper that returns null when BigNumber strict mode would throw (".", "abc", empty), and routed `QuantityInput`'s validate through it — "." now resolves to the "Enter amount" error and Confirm stays disabled.

While auditing, the same `BigNumber(input || 0)` pattern was present in other user-input sites (TransferFields, TransferFooter, BridgeFields, SendFields, bridge/data/simulate) and was equally throw-prone, so all of them were migrated to the helper in the same commit. Verified end-to-end in Playwright across Deposit/Bridge/Send with "." input — no console errors and Confirm remained disabled in each.

File: packages/interwovenkit-react/src/pages/deposit/TransferFields.tsx [hunk 5]
```diff
@@ -147,7 +148,7 @@ const TransferFields = ({ mode }: Props) => {
   const sourceBalance = balances?.[srcChainId]?.[srcDenom]?.amount
   const price = balances?.[srcChainId]?.[srcDenom]?.price
 
-  const quantityValue = BigNumber(price ?? 0).times(rawQuantity || 0)
+  const quantityValue = BigNumber(price ?? 0).times(parseQuantity(rawQuantity) ?? 0)
```
[the remark this answers is not in the document]
Comment (simcheolhwan) [reply packages/interwovenkit-react/src/pages/deposit/TransferFields.tsx]: Switched both fallbacks to `|| 0` in be89bed to satisfy the strict BigNumber fallback rule consistently.

File: packages/interwovenkit-react/src/pages/deposit/TransferFields.tsx [hunk 6]
```diff
@@ -344,7 +345,7 @@ const TransferFields = ({ mode }: Props) => {
             className={styles.maxButton}
             onClick={() => {
               const maxAmount = fromBaseUnit(balance, { decimals: amountDecimals })
-              if (BigNumber(rawQuantity || 0).eq(maxAmount)) return
+              if (parseQuantity(rawQuantity)?.eq(maxAmount)) return
```
[the remark this answers is not in the document]
Comment (simcheolhwan) [reply packages/interwovenkit-react/src/pages/deposit/TransferFields.tsx]: Good catch — added `|| 0` fallbacks to all three `fromBaseUnit` results (lines 180-181, 194-195, 347-348) in be89bed so the downstream `.gt()`/`.eq()` calls stay safe in strict mode.

File: packages/interwovenkit-react/src/components/form/QuantityInput.tsx [hunk 7]
```diff
@@ -1,16 +1,13 @@
...
 import NumericInput from "./NumericInput"
 import styles from "./QuantityInput.module.css"
 
 const QuantityInputReadOnly = ({ children }: { children: string }) => {
-  return (
-    <p className={clsx(styles.input, { [styles.placeholder]: BigNumber(children).isZero() })}>
-      {children}
-    </p>
-  )
+  const parsed = parseQuantity(children)
```
Comment (evilpeach) [inline packages/interwovenkit-react/src/components/form/QuantityInput.tsx]: Since this change makes `QuantityInput.ReadOnly` use `parseQuantity(children)`, formatted bridge outputs can be misclassified as placeholders.

Steps to reproduce:

1. Bridge an amount where the destination display is `>= 1,000`.
2. `BridgeFields.tsx` formats `route.amount_out` with `formatAmount(...)`, which adds comma separators, e.g. `"1,000.000000"`.
3. That formatted string is passed to `QuantityInput.ReadOnly`.
4. `QuantityInput.ReadOnly` calls `parseQuantity(children)`.
5. `parseQuantity("1,000.000000")` returns `null`, so `isPlaceholder` becomes `true`.

In case w/o comma
Image

In case w/ comma
Image

Comment (simcheolhwan) [reply packages/interwovenkit-react/src/components/form/QuantityInput.tsx]: Fixed in f5257aa. `QuantityInput.ReadOnly` now strips commas before calling `parseQuantity`, so `formatAmount`-produced strings like `"1,000.000000"` parse correctly and no longer fall back to the placeholder style.

Verified via Playwright by stubbing `received` to `"1,000.000000"`, `"1,234,567.890123"`, `"0.000000"`, and `"0"` — placeholder class only applies to zero values, exactly as before for the non-formatted cases.

While I was there I audited the rest of the codebase for the same pattern (formatted strings flowing back into BigNumber/parseQuantity) and didn't find any other risky sites: `NumericInput` stores the raw `values.value` from `react-number-format`, the `BRIDGE_QUANTITY` in localStorage is that same raw value, and the remaining `formatAmount` results are only consumed as JSX text.

File: packages/interwovenkit-react/src/pages/bridge/data/fee-warning.ts [hunk 8]
```diff
@@ -26,7 +26,9 @@ export function shouldWarnInsufficientFeeBalance({
   }
 
   const feeRequirementsByDenom = additionalFees.reduce((map, fee) => {
-    if (fee.amount === undefined) return map
+    // Skip API can return `amount: undefined` or `amount: ""` when a per-fee value isn't quoted.
+    // Both are treated as "no contribution"; "" must be filtered explicitly because BigNumber("") throws under strict mode.
+    if (fee.amount === undefined || fee.amount === "") return map
```
[the remark this answers is not in the document]
Comment (simcheolhwan) [reply packages/interwovenkit-react/src/pages/bridge/data/fee-warning.ts:31]: Not acting on this. Skip API's `fee.amount`/`route.amount_in`/`balance.amount` are typed `string` and serialized from numeric values; the realistic non-numeric cases are `undefined` and `""`, which the explicit `=== ""` and `=== undefined` checks already cover. Whitespace-only payloads aren't a documented or observed contract.

The `(amount || "").trim()` in `fee.ts#normalizeL1GasPriceEntry` is not a precedent for handling whitespace amounts — it was added in 57edd03 to address @tansawit's point that `.trim()` itself would throw if the API sent `undefined`/`null`. The intent there is null-safety on the existing call site, not defending against whitespace input.

If we ever see Skip return a whitespace-only amount in production, we can revisit; until then adding `.trim()` here would be speculative defense at the cost of muddier intent.

**Review by evilpeach — APPROVED** [review]
